File: backend/app/models/timeline.py
from datetime import datetime
from typing import Optional, Dict
from sqlmodel import SQLModel, Field, JSON
from uuid import UUID, uuid4

class TimelineEventBase(SQLModel):
    event_type: str  # e.g., "role_change", "discount_applied", "booking_cancelled"
    actor_id: UUID = Field(index=True)
    actor_req_role: str # Role of the actor at the time
    target_id: Optional[str] = Field(default=None, index=True) # ID of user/booking affected
    target_type: str # "user", "booking"
    description: str
    metadata_dump: Dict = Field(default={}, sa_type=JSON) # Stores before/after values
    timestamp: datetime = Field(default_factory=datetime.utcnow)

class TimelineEvent(TimelineEventBase, table=True):
    id: UUID = Field(default_factory=uuid4, primary_key=True)

class TimelineEventCreate(TimelineEventBase):
    pass

class TimelineEventRead(TimelineEventBase):
    id: UUID

File: src/api/timeline.ts
import { api as client } from './client';

export interface TimelineEvent {
    id: string;
    event_type: string;
    actor_id: string;
    actor_req_role: string;
    target_id: string;
    target_type: string;
    description: string;
    metadata_dump: Record<string, any>;
    timestamp: string;
}

export const fetchTimelineEvents = async (params: {
    target_id?: string;
    event_type?: string;
    limit?: number
} = {}) => {
    const response = await client.get<TimelineEvent[]>('/timeline/', { params });
    return response.data;
};
